### DISCUSSION TITLE
SoftConsole projects converted to Cmake

### DISCUSSION THREAD

**Discussion by rockoguido** [polarfire-soc/polarfire-soc-documentation#188] [Bare metal embedded software] [answered]
Hoping this is a proper place to raise this question....

working with a PolarFire SoC user.    on the SoC side and the fabric side.      in the case targeting the MiV in the Fabric trying to help the customer migrate away from SoftConsole for building apps.   They are trying to establish a traditional automated build system     We know that Makefiles are generated via Softconsole and they work fine with GNU Make.   Any work been done using Cmake instead of GNU Make?   Anyone been down this path???

**Comment by AntonKrug** [comment, marked as answer]
I do not know any easy and reliable way to achieve exactly what you ask for. However why it must be cmake, if you want, you can generate makefiles and use them as they are. The makefiles are not cool enough for "traditional automated build system"? Yes cmake is cooler, but for building the stuff does it matter what is used when it will do what you want in the end? Or even better, if you want an automated build system then do not convert SoftConsole projects, keep them as they are and you can use the SoftConsole headless builder, there are even docker containers exactly focused for that: https://hub.docker.com/r/microsemiproess/softconsole-headless-slim/tags

That container can be used locally or in the cloud, manually or automatically, jenkins/travis/github/bitbucket pipelines work too. Often it's just a docker single liner to get it built, for example like here:
https://github.com/RISCV-on-Microsemi-FPGA/SoftConsole/blob/master/.travis.yml

Give it a repository URL it will clone it if it's git one, give it a zip URL, it will download the zip and it will extract and import all included projects. The idea was to get things built/compiled with as few commands as possible.

**Reply by rockoguido** [reply]
Yes Anton you make a great point and I pointed the customer to the use of GNU Make and the fact that he has Makefiles ready to go.  

Let me look at the the use of the headless builder for SoftConsole, that is an interesting option..  
Also looking at the removal of absolute paths in the GNU Makefiles that SoftConsole/Eclipse creates.  It seems this is the default behavior and much has been written about it.
I did notice that in language.settings.yml there are 2 instances of 
keep-relative-paths="false"
I have yet to test as to whether if this setting it set to true does allow relative paths in the Makefiles.

**Reply by AntonKrug** [reply]
Do you mean the include paths? If you are using `${workspace_loc}` macro in the paths, then you are asking the tool to resolve it as a full path, so it's doing what you are asking for. I think if you use relative path in your includes without macros, then the Makefile could have relative paths as well.

Look at the bundled workspace examples, the bundled workspace is there on purpose.

![image](https://user-images.githubusercontent.com/30338596/189112930-207c0d2c-1786-499c-a015-1f947cc0ed82.png)
